File: app/src/main/java/edu/illinois/cs/cs124/ay2024/mp/activities/RSOActivity.java
package edu.illinois.cs.cs124.ay2024.mp.activities;

import static org.junit.Assert.assertNotNull;
import static edu.illinois.cs.cs124.ay2024.mp.models.Summary.filterColor;
//import static edu.illinois.cs.cs124.ay2024.mp.models.Summary.sortByFavorite;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.CompoundButton;
import android.widget.TextView;
import android.widget.ToggleButton;
import androidx.annotation.Nullable;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import edu.illinois.cs.cs124.ay2024.mp.R;
import edu.illinois.cs.cs124.ay2024.mp.adapters.SummaryListAdapter;
import edu.illinois.cs.cs124.ay2024.mp.application.JoinableApplication;
import edu.illinois.cs.cs124.ay2024.mp.models.RSO;
import edu.illinois.cs.cs124.ay2024.mp.models.Summary;
import edu.illinois.cs.cs124.ay2024.mp.network.Client;
import edu.illinois.cs.cs124.ay2024.mp.network.Server;

public class RSOActivity extends Activity {
  /** Tag to identify the RSOActivity in the logs. */
  private static final String TAG = RSOActivity.class.getSimpleName();

  private List<Summary> summaries = Collections.emptyList();
  private final Set<Summary.Color> shownColors = new HashSet<>(Arrays.asList(
      Summary.Color.ORANGE, Summary.Color.BLUE, Summary.Color.DEPARTMENT
  ));

  private SummaryListAdapter listAdapter;
  private String rsoId;
  private ToggleButton favoriteButton;

  private final Map<String, Boolean> favoriteMap = new HashMap<>();

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    // Load this activity's layout and set the title
    setContentView(R.layout.activity_rso);

    rsoId = getIntent().getStringExtra("id");
    Map<String, RSO> map = Server.getIdToRSO();
    RSO rso;
    if (!map.containsKey(rsoId)) {
      // If no RSO found with that ID, just return
      return;
    } else {
      rso = map.get(rsoId);
    }
    assert rso != null;
    String title = rso.getTitle();
    String website = rso.getWebsite();
    String mission = rso.getMission();
    String finalCats = String.join(", ", rso.getCategories());

    TextView titleText = findViewById(R.id.example);
    titleText.setText(title);
    TextView webText = findViewById(R.id.web);
    webText.setText(website);
    TextView missText = findViewById(R.id.miss);
    missText.setText(mission);
    TextView catText = findViewById(R.id.cats);
    catText.setText(finalCats);

    // Initially, create the list adapter with an empty list
    listAdapter = new SummaryListAdapter(summaries, this, summary -> {
      Log.d(TAG, "User clicked on " + summary.getTitle());
      Intent rsoActivity = new Intent(this, RSOActivity.class);
      assertNotNull("Intent component is null", rsoActivity.getComponent());
      rsoActivity.putExtra("id", summary.getId());
      startActivity(rsoActivity);
    });

    favoriteButton = findViewById(R.id.favoriteButton);
    favoriteButton.setOnCheckedChangeListener(favoriteButtonListener);
    Log.d(TAG, "RSOActivity launched");
  }

  @Override
  protected void onStart() {
    super.onStart();

    // Access the application and retrieve the client to get summaries
    JoinableApplication application = (JoinableApplication) getApplication();
    Client client = application.getClient();

    // Retrieve summaries from the server
    client.getSummaries(result -> {
      try {
        List<Summary> allSummaries = result.getValue();
        // Filter by shownColors if needed
        summaries = filterColor(allSummaries, shownColors);
        Collections.sort(summaries);
        runOnUiThread(() -> {
          listAdapter.setSummaries(summaries);
          listAdapter.notifyDataSetChanged();
        });
      } catch (Exception e) {
        Log.e(TAG, "Error updating summary list", e);
      }
    });

    // Retrieve the current favorite status of this RSO
    client.getFavorite(rsoId, favoriteResult -> {
      if (!(favoriteResult == null)) {
        boolean isFavorite = favoriteResult.getValue();
        runOnUiThread(() -> favoriteButton.setChecked(isFavorite));
      }
    });
  }

  private final CompoundButton.OnCheckedChangeListener favoriteButtonListener = (unused, checked) -> {
    Log.d(TAG, "Favorite button checked " + checked);
    JoinableApplication application = (JoinableApplication) getApplication();
    Client client = application.getClient();
    // Set the favorite status for the current RSO based on the button state
    client.setFavorite(rsoId, checked, result -> {
      runOnUiThread(() -> {
        listAdapter.notifyDataSetChanged();
      });
      Log.d(TAG, "Favorite status set to: " + checked);
    });
  };
}

File: app/src/main/java/edu/illinois/cs/cs124/ay2024/mp/activities/RSOActivity.kt
package edu.illinois.cs.cs124.ay2024.mp.activities

import android.app.Activity
import android.os.Bundle
import android.util.Log
import android.widget.TextView
import android.widget.ToggleButton
import edu.illinois.cs.cs124.ay2024.mp.R
import edu.illinois.cs.cs124.ay2024.mp.application.JoinableApplication
import edu.illinois.cs.cs124.ay2024.mp.network.Server

class RSOActivity : Activity() {
    private val TAG = RSOActivity::class.java.simpleName
    private lateinit var rsoId: String
    private lateinit var favoriteButton: ToggleButton

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_rso)

        rsoId = intent.getStringExtra("id") ?: return
        val rso = Server.getIdToRSO()[rsoId] ?: return

        findViewById<TextView>(R.id.example).text = rso.title
        findViewById<TextView>(R.id.web).text = rso.website
        findViewById<TextView>(R.id.miss).text = rso.mission
        findViewById<TextView>(R.id.cats).text = rso.categories.joinToString(", ")

        favoriteButton = findViewById(R.id.favoriteButton)
        favoriteButton.setOnCheckedChangeListener { _, checked ->
            Log.d(TAG, "Favorite button checked $checked")
            val client = (application as JoinableApplication).getClient()
            client.setFavorite(rsoId, checked) { _ ->
                runOnUiThread { finish() }
                Log.d(TAG, "Favorite status set to: $checked")
            }
        }
    }

    override fun onStart() {
        super.onStart()
        if (!::rsoId.isInitialized) return
        val client = (application as JoinableApplication).getClient()
        client.getFavorite(rsoId) { result ->
            if (result != null) {
                val isFavorite = result.getValue()
                runOnUiThread { favoriteButton.isChecked = isFavorite }
            }
        }
    }
}
